File: python/S14_BST/L1_CONCEPT/2_searchbst.py
class Node:
    def __init__(self,data):
        self.left=None
        self.right=None
        self.data=data
class bst:
    def __init__(self):
        self.root=None
    def insertbst(self,root,data):
        if root is None:
            return Node(data)
        if root.data>data:
            root.left=self.insertbst(root.left,data)
        elif root.data<data:
            root.right=self.insertbst(root.right,data)
        return root
        
            
    def search(self,root,key):
        if root is None:
            return None
        if root.data==key:
            return root
        if root.data>key:
            return self.search(root.left,key)
        elif root.data<key:
            return self.search(root.right,key)
        
    
    def postorder(self,root):
        if root is None:
            return
        self.postorder(root.left)
        self.postorder(root.right)
        print(root.data,end=" ")
if __name__=="__main__":
    b=bst()
    b.root=b.insertbst(b.root,34)
    b.root=b.insertbst(b.root,30)
    b.root=b.insertbst(b.root,35)
    b.root=b.insertbst(b.root,33)
    b.root=b.insertbst(b.root,32)
    temp=b.search(b.root,34)
    if temp is None:
        print("no")
    else:
        print("yes")
    b.postorder(b.root)
    

File: C/S14_BST/L1_CONCEPT/2_searchbst.c
#include<stdio.h>
#include<stdlib.h>
struct node{
    int data;
    struct node *left;
    struct node *right;
};
struct node *createnode(int data){
    struct node *root=(struct node*)malloc(sizeof(struct node));
    root->left=NULL;
    root->right=NULL;
    root->data=data;
    return root;
}
struct node *insertbst(struct node *root,int data){
    struct node *newnode=createnode(data);
    if(root==NULL){
        return newnode;
    }
    if(root->data<data){
        root->right=insertbst(root->right,data);
    }
    else if(root->data>data){
        root->left=insertbst(root->left,data);
    }
    return root;
}
int search(struct node *root,int key){
    if(root==NULL)return 0;
    if(root->data==key){
        return 1;
    }
    if(root->data>key){
        return search(root->left,key);
    }
    return search(root->right,key);
}
void postorder(struct node *root){
    if(root==NULL)return;
    postorder(root->left);
    postorder(root->right);
    printf("%d ",root->data);
}
int main(){
    struct node *root=NULL;
    root=insertbst(root,56);
    root=insertbst(root,564);
    int s=search(root,596);
    postorder(root);
    printf("\n%d",s);
    return 0;
}